File: artifacts/discord-bot/bot/services/warningService.js
const db = require('../database/sqlite');

/**
 * Add a warning for a user in a guild.
 */
async function addWarning(userId, guildId, moderatorId, reason) {
  const timestamp = new Date().toISOString();
  const info = db.prepare(`
    INSERT INTO warnings (user_id, guild_id, moderator_id, reason, timestamp)
    VALUES (?, ?, ?, ?, ?)
  `).run(userId, guildId, moderatorId, reason, timestamp);
  
  return { id: info.lastInsertRowid, user_id: userId, guild_id: guildId, moderator_id: moderatorId, reason, timestamp };
}

/**
 * Get all warnings for a user in a guild.
 */
async function getWarnings(userId, guildId) {
  return db.prepare(`
    SELECT * FROM warnings 
    WHERE user_id = ? AND guild_id = ? 
    ORDER BY timestamp DESC
  `).all(userId, guildId);
}

/**
 * Get warning count for a user in a guild.
 */
async function getWarningCount(userId, guildId) {
  const row = db.prepare(`
    SELECT COUNT(*) as count FROM warnings 
    WHERE user_id = ? AND guild_id = ?
  `).get(userId, guildId);
  return row ? row.count : 0;
}

/**
 * Clear all warnings for a user in a guild.
 */
async function clearWarnings(userId, guildId) {
  db.prepare(`
    DELETE FROM warnings 
    WHERE user_id = ? AND guild_id = ?
  `).run(userId, guildId);
}

module.exports = { addWarning, getWarnings, getWarningCount, clearWarnings };

File: services/warning_service.py
from datetime import datetime
from database.connection import get_db

async def add_warning(guild_id: str | int, user_id: str | int, moderator_id: str | int, reason: str, warn_type: str = "warn") -> int:
    gid, uid, mid = str(guild_id), str(user_id), str(moderator_id)
    now = datetime.utcnow().isoformat()
    async with get_db() as db:
        cursor = await db.execute(
            """
            INSERT INTO warnings (guild_id, user_id, moderator_id, reason, type, timestamp)
            VALUES (?, ?, ?, ?, ?, ?)
            """,
            (gid, uid, mid, reason, warn_type, now)
        )
        await db.commit()
        return cursor.lastrowid

async def get_warnings(guild_id: str | int, user_id: str | int) -> list[dict]:
    gid, uid = str(guild_id), str(user_id)
    async with get_db() as db:
        async with db.execute(
            "SELECT * FROM warnings WHERE guild_id = ? AND user_id = ? ORDER BY id DESC",
            (gid, uid)
        ) as cursor:
            rows = await cursor.fetchall()
            return [dict(r) for r in rows]

async def clear_warnings(guild_id: str | int, user_id: str | int) -> int:
    gid, uid = str(guild_id), str(user_id)
    async with get_db() as db:
        cursor = await db.execute(
            "DELETE FROM warnings WHERE guild_id = ? AND user_id = ?",
            (gid, uid)
        )
        await db.commit()
        return cursor.rowcount
